vectordraw/grader.py

```diff
@@ -0,0 +1,381 @@
+        return _errmsg('The angle of {name} is incorrect. Your angle: {angle:.1f}', check, vectors)
+
+
+def check_segment_angle(check, vectors):
+    """
+    Check if angle of segment targeted by `check` is correct.
+    """
+    # Segments are not directed, so we must check the angle between the segment and
+    # the vector that represents it, as well as its opposite vector.
+    vec = vectors[check['vector']]
+    tolerance = check.get('tolerance', 2.0)
+    expected = math.radians(check['expected'])
+    if not (_angle_within_tolerance(vec, expected, tolerance) or
+            _angle_within_tolerance(vec.opposite(), expected, tolerance)):
+        return _errmsg('The angle of {name} is incorrect. Your angle: {angle:.1f}', check, vectors)
+
+
+def _dist_line_point(line, point):
+    """
+    Return distance between `line` and `point`.
+
+    The line is passed in as a Vector instance, the point as a Point instance.
+    """
+    direction_x = line.tip.x - line.tail.x
+    direction_y = line.tip.y - line.tail.y
+    determinant = (point.x - line.tail.x) * direction_y - (point.y - line.tail.y) * direction_x
+    return abs(determinant) / math.hypot(direction_x, direction_y)
+
+
+def check_points_on_line(check, vectors):
+    """
+    Check if line targeted by `check` passes through correct points.
+    """
+    line = vectors[check['vector']]
+    tolerance = check.get('tolerance', 1.0)
+    points = check['expected']
+    for point in points:
+        point = Point(point[0], point[1])
+        if _dist_line_point(line, point) > tolerance:
+            return _errmsg(
+                'The line {name} does not pass through the correct points.', check, vectors
+            )
+
+
+def check_point_coords(check, points):
+    """
+    Check if coordinates of point targeted by `check` are correct.
+    """
+    point = points[check['point']]
+    tolerance = check.get('tolerance', 1.0)
+    expected = check['expected']
+    dist = math.hypot(expected[0] - point.x, expected[1] - point.y)
+    if dist > tolerance:
+        return _errmsg_point('Point {name} is not at the correct location.', check, point)
+
+
+class Point(object):
+    """ Represents a single point on the vector drawing board. """
+    def __init__(self, x, y):
+        self.x = x
+        self.y = y
+
+
+class Vector(object):
+    """ Represents a single vector on the vector drawing board. """
+    def __init__(self, name, x1, y1, x2, y2):  # pylint: disable=too-many-arguments
+        self.name = name
+        self.tail = Point(x1, y1)
+        self.tip = Point(x2, y2)
+        self.length = math.hypot(x2 - x1, y2 - y1)
+        angle = math.degrees(math.atan2(y2 - y1, x2 - x1))
+        if angle < 0:
+            angle += 360
+        self.angle = angle
+
+    def opposite(self):
+        """
+        Return new vector with tip and tail swapped.
+        """
+        return Vector(self.name, self.tip.x, self.tip.y, self.tail.x, self.tail.y)
+
+
+class Grader(object):
```

Comment:
@singingwolfboy You're right about the fact that the `grade` method is the only thing being called from outside. But I think it makes sense to encapsulate the grading logic in a class, because otherwise we'd end up with either one long function that has all of the logic from the `Grader` class, or multiple functions that you wouldn't know are related just by looking at them (i.e., you'd have to read the code in more detail to understand the relationship between the functions).